### PR TITLE
Fix menu redraw for DOGLCD, improve MBL

### PR DESCRIPTION
**Pull request by thinkyhead**
- Do extra redraws on Manual Bed Leveling screens for graphical displays
- Fix "jumping" move bug by resetting `encoderPosition` even if no move is processed
- Check `axis_homed` instead of `axis_known_position` in `_lcd_level_bed_homing`
- Show a "`Next Point: 1 / 9`" display while moving to each probe point
- Break up Manual Bed Leveling into more menu handlers
- Software endstops don't matter for MBL Z adjustment, just use the given range
- The `mbl_wait_for_move` variable is dropped. Use this procedure instead:
  - Before some blocking move call `lcd_goto_menu` with the screen you want to display
  - During `st_synchronize` that screen will be displayed
  - After the blocking function call `lcd_goto_menu` with the next screen to display
  - Note that `_lcd_level_bed_homing` simply monitors itself

The main issue that can cause `lcdDrawUpdate` to have a harder time with graphical displays is that each menu handler is called two or more times for a graphical display. The first time the handler is called, `lcdDrawUpdate` will be what the previous setup code (from reading the controller, or from the previous handler) intended. If the handler sets `lcdDrawUpdate` to zero, then the next call won't draw.

For graphical displays the menu-screen draw and process loop ought to pass a flag to the current menu handler that it should only draw and not process (i.e., After the first call in the loop). Currently this is worked around by (a) ignoring a repeated click on the LCD, and (b) setting `encoderPosition` to zero after processing it. These extra steps ensure that only drawing will take place after the first controller event.

All this being said, I haven't found any places where the current arrangement is a problem.

Comparing the older assumptions with the newest ones, it's clear that the purpose of setting `lcdDrawUpdate = 1` within handlers (in the old implementation) was to make sure the screen would be drawn on the next call in the graphical display loop (if there was one). So it makes sense that it's only cleared after the "screen draw loop" is complete, and before the next "menu event loop."

The new system is cleaner, but up to now it failed to consider what was being done to help the graphical display loop work. This PR is working well now, but I should go through the menu system some more to make sure it's being properly friendly to graphical displays in the way I've outlined.

References #3533, AnHardt/Marlin#30

### CODE REVIEW

File: Marlin/ultralcd.cpp [hunk 1]
```diff
@@ -914,114 +912,154 @@ void lcd_cooldown() {
...
-    lcdDrawUpdate = LCDVIEW_CALL_NO_REDRAW;
+    lcdDrawUpdate =
+      #if ENABLED(DOGLCD)
+        LCDVIEW_CALL_REDRAW_NEXT
+      #else
+        LCDVIEW_CALL_NO_REDRAW
+      #endif
+    ;
     if (axis_known_position[X_AXIS] && axis_known_position[Y_AXIS] && axis_known_position[Z_AXIS])
       lcd_goto_menu(_lcd_level_bed_homing_done);
```
Comment (Blue-Marlin) [inline Marlin/ultralcd.cpp:1025]: The `axis_known_position` vs `axis_homed` glitch is still in.
https://github.com/MarlinFirmware/Marlin/issues/3533#issuecomment-212509170

Comment (thinkyhead) [inline Marlin/ultralcd.cpp:1025]: Aha!

File: Marlin/ultralcd.cpp [hunk 2]
```diff
@@ -914,125 +912,165 @@ void lcd_cooldown() {
...
+  static void _lcd_level_goto_next_point();
+
+  static void _lcd_level_bed_done() {
+    if (lcdDrawUpdate) lcd_implementation_drawedit(PSTR(MSG_LEVEL_BED_DONE));
+    lcdDrawUpdate =
+      #if ENABLED(DOGLCD)
+        LCDVIEW_CALL_REDRAW_NEXT
+      #else
+        LCDVIEW_CALL_NO_REDRAW
+      #endif
```
Comment (thinkyhead) [inline Marlin/ultralcd.cpp]: This is the kind of change I did to make sure graphical displays get a complete update. But it's a workaround for the "bad idea" it started with. The value of `lcdDrawUpdate` shouldn't be changed _between graphical draw calls_. So it probably works to simply not change `lcdDrawUpdate` here at all!

File: Marlin/ultralcd.cpp [hunk 3]
```diff
@@ -914,125 +912,165 @@ void lcd_cooldown() {
...
-    if (lcdDrawUpdate) {
-      float v = current_position[Z_AXIS] - MESH_HOME_SEARCH_Z;
-      lcd_implementation_drawedit(PSTR(MSG_MOVE_Z), ftostr43(v + (v < 0 ? -0.0001 : 0.0001), '+'));
+      lcdDrawUpdate =
+        #if ENABLED(DOGLCD)
+          LCDVIEW_CALL_REDRAW_NEXT
+        #else
+          LCDVIEW_REDRAW_NOW
+        #endif
+      ;
```
Comment (thinkyhead) [inline Marlin/ultralcd.cpp]: Here `lcdDrawUpdate = LCDVIEW_REDRAW_NOW` should work fine, as long as `lcdDrawUpdate` isn't changed later. (This function only needs to be called for encoder events and to handle the draw flag set here.)

File: Marlin/ultralcd.cpp [hunk 4]
```diff
@@ -914,125 +912,165 @@ void lcd_cooldown() {
...
+      lcd_implementation_drawedit(PSTR(MSG_LEVEL_BED_NEXT_POINT), msg);
+    }
+
+    lcdDrawUpdate =
+      #if ENABLED(DOGLCD)
+        LCDVIEW_CALL_REDRAW_NEXT
+      #else
+        LCDVIEW_CALL_NO_REDRAW
+      #endif
+    ;
```
Comment (thinkyhead) [inline Marlin/ultralcd.cpp]: No callback is needed, and so this should work if `lcdDrawUpdate` is simply left unchanged.

File: Marlin/ultralcd.cpp [hunk 5]
```diff
@@ -914,125 +912,165 @@ void lcd_cooldown() {
...
     if (lcdDrawUpdate) lcd_implementation_drawedit(PSTR(MSG_LEVEL_BED_HOMING), NULL);
-    lcdDrawUpdate = LCDVIEW_CALL_NO_REDRAW;
-    if (axis_known_position[X_AXIS] && axis_known_position[Y_AXIS] && axis_known_position[Z_AXIS])
+    lcdDrawUpdate =
+      #if ENABLED(DOGLCD)
+        LCDVIEW_CALL_REDRAW_NEXT
+      #else
+        LCDVIEW_CALL_NO_REDRAW
+      #endif
+    ;
```
Comment (thinkyhead) [inline Marlin/ultralcd.cpp]: This is a little tricky too. We want this to draw the first time through (for all graphical display updates) but then not redraw again. However we still want the callback.

Another issue that happens (and the reason that the bottom is drawn but not the top) is that `lcd_goto_menu` takes effect right away, so in the graphical display loop only the second half of the next screen will be drawn. Thus `lcd_goto_menu` should use a temporary pointer and only change the real `currentMenu` pointer after the drawing loop.

In fact, both of these items can be solved using a `next_thing_value` that supplants the old one, only applying after the drawing loop.

**Comment by thinkyhead** [comment]
Ok, I think I got it after myriad tweaks.

- Don't switch menu pages in the middle of a display update
- Don't change `lcdDrawUpdate` in a menu handler (unless a value changes without an encoder event)

To ensure the first, switching menu pages is now deferred until after the draw loop.

To ensure the second, I removed almost all changes to `lcdDrawUpdate` except to get the forced callback in `_lcd_bed_level_homing`.

I haven't been able to test these latest updates with my LCD but will do so as soon as I get home!

**Comment by AnHardt** [comment]
What you are telling us, looks as if you understood the u8glig display loop and its implications now (https://github.com/olikraus/u8glib/wiki/tpictureloop). Congratulations.

Alone this PR makes the display part for MBL worse than before.
- Missing upper half when displaying "Homing XYZ".
- No display after clicked "Click to Begin" before an 'event' occurs.
- No "Next Point: 1 / 9" displays at all.
- When setting a z, the display stays on the old screen for about a second (could be the duration of the move) to display an empty screen then, until an event occurs.

**Comment by garaguido** [comment]
Hi , I used the latest " RCBugfix  6fac4d9 "from 21-04-16 en uploaded it , the LCD behave weird also there where no digits on the display for adjust Z hight  , also de Y steppermotor  has a verry slow behavier after MBL and Homing X Y Z . 
I have used again the older version from 20-04-16 with the mods from Bleu Marlin its working fine by me

**Comment by AnHardt** [comment]
A try for an analysis:
Let's begin in `lcd_level_bed()`. This is a normal menu. We select "Level Be" and click. We selected a submenu, so nextMenu is set to `_lcd_level_bed_continue`. The END_MENU() sets `lcdDrawUpdate = LCDVIEW_CALL_REDRAW_NEXT;` So when we leave the display loop we are currently in `currentmenu` is set to `_lcd_level_bed_continue` and `lcdDrawUpdate` is set to `LCDVIEW_REDRAW_NOW`. 
A moment later we are calling `currentmenu()` again. that means  `_lcd_level_bed_continue()`. We do make some preparations, enqueue G28 and order `_lcd_level_bed_homing` in `nextMenu`.
The second entrance here enques an other G28 (have a look at the serial output).
Leaving the loop again because of `LCDVIEW_REDRAW_NOW` we change `lcdDrawUpdate` to `LCDVIEW_NONE`.
The next `idle()` begins to perform G28. The next call to `lcd_update` sees `LCDVIEW_NONE`.

I suppose, to get a single run of `_lcd_level_bed_continue()` we need a 

```
currentMenu = _lcd_level_bed_continue;
_lcd_level_bed_continue();
```

instead of 

```
lcd_goto_menu(_lcd_level_bed_homing);
```

In `_lcd_level_bed_homing()`, `if (lcdDrawUpdate)` copletely does not make any sense. We wont be here if `lcdDrawUpdate` wont be different from 0.

Sorry. Running out of time for now.

**Comment by epatel** [comment]
This makes me sad. I made MBL over a year ago and feel some responsibility for it. I also have a `REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER`, and it worked/works fine for me. The branch I use myself (which I feel comfortable with) have a little extra http://www.thingiverse.com/thing:1120142
I have not decided yet if I should jump in and try to fix this, it really feels like someone else changed something that changed everything.

File: Marlin/ultralcd.cpp (hunk 5 above)
Comment (epatel) [inline Marlin/ultralcd.cpp]: Why did you remove the `if (..)` statement? If I remember correctly and that is the place I am thinking of that `if (..)` is checking for when the "homing" is done.

**Comment by Blue-Marlin** [comment]
@epatel 
A bit further down it's replaced with `axis_homed`. `axis_known_position` is now an indicator for 'lost precision'. In #3533 the user had set `#define DISABLE_Z true` and got stuck in waiting for the homing to finish.

**Comment by thinkyhead** [comment]
> it really feels like someone else changed something that changed everything

For my part, I haven't been trying to "change" anything, only to fix the display problems that I could see on my character-based (and now graphical) display. If you go back to RC5 and run Manual Bed Leveling you will see that it continually clears of the screen during the "`Homing XYZ`" message, for example. That's because the original implementation of MBL was hacked onto a menu system that didn't provide a way to have a callback without clearing the screen. So I've been gradually filling in holes in the menu system so that it's possible to have Manual Bed Leveling work as it was obviously designed. I can only apologize that Marlin was not yet ready for it.

Once these display glitches with the menu system are all worked out, I think you will find that Manual Bed  works much better overall. It's nice to see these messages not flashing wildly, and to have it displaying "`Next Point: 2 / 9`" while the nozzle is moving between points….

**Comment by thinkyhead** [comment]
> What you are telling us, looks as if you understood the u8glig display loop and its implications now (https://github.com/olikraus/u8glib/wiki/tpictureloop). Congratulations.

Thank you! Congratulations on being way ahead of me, you freakin' genius.

**Comment by epatel** [comment]
> If you go back to RC5 and run Manual Bed Leveling you will see that it continually clears of the screen

Thing is I am running an older version, closer to my original commit. Which works perfectly well without flickering, and I am happy with that version. Which also contain an auto- part, which I am also happy with. The commit/branch and a video clip of it running can be accessed/seen from here http://www.thingiverse.com/thing:1120142

**Comment by thinkyhead** [comment]
> I am running an older version

How much older? What are the differences between that version and `RCBugFix`?

> Which works perfectly well without flickering, and I am happy with that version.

Are you using a character display or a graphical display?
